### ISSUE TITLE
output format, what to do when xml is needed

### ISSUE THREAD

**Issue by DHatziioanou** [bbuchfink/diamond#49]
Hello,

While diamond blast is wonderfully fast some applications need blast results in xml format. I haven't been able to locate a tool or method to convert the diamond daa or tabular output to xml, could you include an option to convert the daa files to xml or point me in the direction of a script/tool which can do the job?

Thanks,
Diane

**Comment by bbuchfink** [comment]
Hi Diane,

I plan to add support for the xml format in an upcoming version. It's probably going to take a couple of days.

Benjamin

**Comment by bbuchfink** [comment]
The latest version now supports the xml format.

**Comment by DHatziioanou** [comment]
Hi Benjamin,

Thanks for including support for xml in the new version!
I am having some downstream problems however. I tried feeding the xml output to STM and I got some  errors:
  File "/home/diane/scripts/STM/mapBlastHit.py", line 96, in <module>
    for record in NCBIXML.parse(blastHandle):
  File "/usr/lib/python2.7/dist-packages/Bio/Blast/NCBIXML.py", line 637, in parse
    expat_parser.Parse(text, False)
  File "/usr/lib/python2.7/dist-packages/Bio/Blast/NCBIXML.py", line 109, in endElement
    eval("self.%s()" % method)
  File "<string>", line 1, in <module>
  File "/usr/lib/python2.7/dist-packages/Bio/Blast/NCBIXML.py", line 284, in _end_BlastOutput_query_len
    self._header.query_letters = int(self._value)
ValueError: invalid literal for int() with base 10: ''

When I tried the same dataset and commands using blastx from the blastall v2.2.26 everything worked smoothly so it seems the xml format you use is different from that of NCBI?!?

**Comment by bbuchfink** [comment]
I'm not familiar with STM, can you tell me where to download that?

**Comment by DHatziioanou** [comment]
Hi Benjamin,

You can donwload STM (Scaffolding using Translation Mapping ) and its publication from here: http://www.surget-groba.ch/soft.php

**Comment by bbuchfink** [comment]
Hi Diane,

please try again with the latest version.

**Comment by DHatziioanou** [comment]
Hi Benjamin,

I'm afraid its still got issues:

Traceback (most recent call last):
  File "/home/diane/scripts/STM/mapBlastHit.py", line 96, in <module>
    for record in NCBIXML.parse(blastHandle):
  File "/usr/lib/python2.7/dist-packages/Bio/Blast/NCBIXML.py", line 657, in parse
    expat_parser.Parse(text, False)
xml.parsers.expat.ExpatError: not well-formed (invalid token): line 600, column 121

**Comment by bbuchfink** [comment]
I tried to run STM myself but it worked for me. How big is your query data and reference db, could you make it available to me?

**Comment by DHatziioanou** [comment]
Hi Benjamin, 

It seems your xml format is fine and it was indeed a file size issue with STM, sorry about that!

For your reference:
The blastx reference I used with diamond v0.8.9.71 as a db was the April 2016 version of nr, available at ftp://ftp.ncbi.nlm.nih.gov/blast/db/FASTA/nr.gz whih is 56.3 Gb and 57.6 Gb after formatting for diamond.
The query data was a transcriptome assembly totalling 126557 conigs with a total wize of 155.7 Mb.
The diamond xml file was 1.3 Gb and had 576972 pairwise alignments and 770835 HSSPs covering 59676 queries (I used k10 so I could use the file for other purposes as well).
 I also tried with a diamond xml file with k1 set which was 174.8 Mb with  59669 reported pairwise alignments and 81676 HSSPs for 59669 queries and got almost the same errors - the last error was on line 171 column 12 instead of line 600 column 121.
Traceback (most recent call last):
  File "/home/diane/scripts/STM/mapBlastHit.py", line 96, in <module>
    for record in NCBIXML.parse(blastHandle):
  File "/usr/lib/python2.7/dist-packages/Bio/Blast/NCBIXML.py", line 657, in parse
    expat_parser.Parse(text, False)
xml.parsers.expat.ExpatError: not well-formed (invalid token): line 171, column 12

I then tested it on the swissprot database which is considerably smaller at 265.5 Mb and got 46026 pairwise alignments reported (lower evalue) and 59806 HSSPs from 46026 queries in a 126.4 Mb xml file. This xml was parsed successfully and everything ran smoothly ever after!

Thanks for the good work,
Diane
